rust/src/pgsql/logger.rs

```diff
@@ -0,0 +1,438 @@
+            }
+            PgsqlErrorNoticeFieldType::InternalPosition => {
+                jb.set_string_from_bytes("internal_position", &field.field_value)?;
+            }
+            PgsqlErrorNoticeFieldType::InternalQuery => {
+                jb.set_string_from_bytes("internal_query", &field.field_value)?;
+            }
+            PgsqlErrorNoticeFieldType::Where => {
+                jb.set_string_from_bytes("where", &field.field_value)?;
+            }
+            PgsqlErrorNoticeFieldType::SchemaName => {
+                jb.set_string_from_bytes("schema_name", &field.field_value)?;
+            }
+            PgsqlErrorNoticeFieldType::TableName => {
+                jb.set_string_from_bytes("table_name", &field.field_value)?;
+            }
+            PgsqlErrorNoticeFieldType::ColumnName => {
+                jb.set_string_from_bytes("column_name", &field.field_value)?;
+            }
+            PgsqlErrorNoticeFieldType::DataType => {
+                jb.set_string_from_bytes("data_type", &field.field_value)?;
+            }
+            PgsqlErrorNoticeFieldType::ConstraintName => {
+                jb.set_string_from_bytes("constraint_name", &field.field_value)?;
+            }
+            PgsqlErrorNoticeFieldType::File => {
+                jb.set_string_from_bytes("file", &field.field_value)?;
+            }
+            PgsqlErrorNoticeFieldType::Line => {
+                jb.set_string_from_bytes("line", &field.field_value)?;
+            }
+            PgsqlErrorNoticeFieldType::Routine => {
+                jb.set_string_from_bytes("routine", &field.field_value)?;
+            }
+            PgsqlErrorNoticeFieldType::TerminatorToken => {
+                // do nothing, as that's the terminator token ?
+            }
+            PgsqlErrorNoticeFieldType::UnknownFieldType => {
+                // TODO Question - what should be done here?
+            }
+        }
+    }
+    jb.close()?;
+    Ok(jb)
+}
+
+fn log_pgsql_parameters(params: &PgsqlStartupParameters) -> Result<JsonBuilder, JsonError> {
+    let mut jb = JsonBuilder::new_object();
+    jb.set_string_from_bytes("user", &params.user.value)?;
+    if let Some(PgsqlParameter { name: _, value }) = &params.database {
+        jb.set_string_from_bytes("database", value)?;
+    }
+    // TODO Leaving these out in the less verbose version, and only show them in case extended and file logging is enabled in yaml?
+    // if let Some(vec) = &params.optional_params {
+    //     for param in vec {
+    //         let name = String::from_utf8_lossy(&param.name);
+    //         jb.set_string_from_bytes(&name, &param.value)?;
+    //     }
+    // }
+    jb.close()?;
+    Ok(jb)
+}
+
+#[no_mangle]
+pub extern "C" fn rs_pgsql_logger_log(tx: *mut std::os::raw::c_void, js: &mut JsonBuilder) -> bool {
+    let tx_safe: &mut PgsqlTransaction;
+    unsafe {
+        tx_safe = cast_pointer!(tx, PgsqlTransaction);
+    }
```

Comment:
This whole function should be made unsafe now.  Sometime after you started this we moved all functions that do a raw pointer cast to unsafe.